### PR TITLE
webapi: Add session support, encapsulate logic

### PR DESCRIPTION
**Pull request by rvolosatovs**
@HaicoDorenbos please look at this branch for the API changes.

### CODE REVIEW

File: pkg/webapi/webapi.go [hunk 1]
```diff
@@ -108,204 +142,318 @@ func requestLogger(r *http.Request) *zap.Logger {
...
+	defer wsConn.Close()
 
-		logger.Debug("Retrieving a connection from pool...")
-		trcConn, err := pool.Conn()
-		if err != nil {
-			wsError(wsConn, logger, errors.Wrap(err, "failed to establish connection to TRC"), websocket.CloseInternalServerErr)
-			return
-		}
+	wsConn.EnableWriteCompression(true)
+	wsConn.SetCompressionLevel(flate.BestCompression)
```
Comment (golangci) [inline pkg/webapi/webapi.go]: Error return value of `wsConn.SetCompressionLevel` is not checked

File: cmd/srrs/main.go [hunk 2]
```diff
@@ -155,29 +150,59 @@ func main() {
...
+		mux := http.DefaultServeMux
 
+		webapi.RegisterHandlers(pool, mux)
 		if *static != "" {
-			http.Handle("/", http.FileServer(http.Dir(*static)))
+			mux.Handle("/", http.FileServer(http.Dir(*static)))
 		}
 
-		logger := logger.With(zap.String("listen_addr_tcp", *tcpAddr))
+		// unsecure server
```
Comment (HaicoDorenbos) [inline cmd/srrs/main.go]: Why are you using it, when it is an unsecure server?

File: certificates/key.pem [hunk 3]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (HaicoDorenbos) [inline certificates/key.pem]: Which size is this private RSA key? 2048 bits? Keep in mind that according to predictions RSA keys of 2048 bits will be cracked after year 2048 (https://crypto.stackexchange.com/questions/1978/how-big-an-rsa-key-is-considered-secure-today). If it is 4096 bits then it is no problem...

File: cmd/srrs/main.go [hunk 4]
```diff
@@ -155,29 +150,59 @@ func main() {
...
+				tcpErrCh <- errors.Wrap(err, "failed to listen")
+			}
+		}()
+
+		tlsErrCh := make(chan error, 1)
+		go func() {
+			tcpLogger.Info("Starting the secure web server...")
+			cert := filepath.Join("certificates", "cert.pem")
+			key := filepath.Join("certificates", "key.pem")
+			if err := tlsSrv.ListenAndServeTLS(cert, key); err != nil {
```
Comment (HaicoDorenbos) [inline cmd/srrs/main.go]: Good job 👍 because of this Man in the Middle Attacks are not possible

File: pkg/api/api.go [hunk 5]
```diff
@@ -8,6 +9,7 @@ import (
 	"github.com/oklog/ulid"
 )
 
+// Command is a TRC command.
```
Comment (HaicoDorenbos) [inline pkg/api/api.go]: Good job 👍 adding comments to variables will make it more readable. But don't forget to add comments to the WebAPI.go file as well

File: cmd/srrs/main_test.go [hunk 6]
```diff
@@ -91,32 +109,46 @@ func TestMain(m *testing.M) {
...
 		defer wg.Done()
 
-		wsAddr := "ws://localhost" + defaultAddr + "/" + stateEndpoint
-		logger.With("addr", wsAddr).Debug("Opening a WebSocket...")
-		wsConn, _, wsErr = websocket.DefaultDialer.Dial(wsAddr, http.Header{
-			"Authorization": []string{fmt.Sprintf("Basic %s", base64.StdEncoding.EncodeToString(append([]byte("user:"), []byte(hs.Token)...)))},
-		})
+		req, err := http.NewRequest(http.MethodGet, "http://"+defaultAddr+"/"+webapi.AuthEndpoint, nil)
+		a.NoError(err)
+		req.SetBasicAuth("", handshake.Token)
```
Comment (HaicoDorenbos) [inline cmd/srrs/main_test.go]: Is this no problem for Google Chrome?

File: cmd/srrs/main.go (hunk 2 above)
Comment (rvolosatovs) [reply cmd/srrs/main.go]: we need it for development/testing, but also we probably want to fallback

**Comment by rvolosatovs** [comment]
@ieperen3039 please open a separate pull request. TLS does not belong here.
I'll force push

File: cmd/srrs/main_test.go (hunk 6 above)
Comment (rvolosatovs) [reply cmd/srrs/main_test.go]: no

File: pkg/api/api.go (hunk 5 above)
Comment (rvolosatovs) [reply pkg/api/api.go]: Everything exported by `pkg/webapi` is commented.

**Comment by rvolosatovs** [comment]
@Huubvr please put your type-related changes in this PR, currently the tests fail, because they expect the diff to be sent, however that is not possible with the current implementation.

**Comment by helaan** [comment]
Disagree, I think we should do the diff-related changes in a different PR, then merge that PR before this one. This pull request is already getting quite large, I don't want another feature/73-ui branch.

**Comment by rvolosatovs** [comment]
@helaan you're not supposed to be reviewing this anyway :slightly_smiling_face:

File: pkg/api/apitest/apitest.go [hunk 7]
```diff
@@ -12,45 +12,274 @@ import (
...
 
-	var pld interface{}
-	switch msg.Type {
-	case api.MessageTypeHandshake:
-		pld = *RandomHandshake()
-	case api.MessageTypeState:
-		pld = *RandomState()
-	case api.MessageTypePing:
-		pld = nil
+func must(v interface{}, err error) interface{} {
```
Comment (golangci) [inline pkg/api/apitest/apitest.go]: `must` is unused

File: cmd/srrs/main_test.go [hunk 8]
```diff
@@ -90,33 +115,49 @@ func TestMain(m *testing.M) {
 	os.Exit(ret)
 }
 
-func TestAll(t *testing.T) {
-	a := require.New(t)
+func TestAPI(t *testing.T) {
```
Comment (golangci) [inline cmd/srrs/main_test.go]: `TestAPI` is unused

**Pull request merged**